File: packages/shared/src/types/progressCounter.ts
/**
 * ProgressCounter - Shared cumulative counter across boards
 *
 * Enables cross-board goals like "Read 100 pages total across 3 boards"
 *
 * Design principles:
 * - Shared across all boards for a user
 * - Atomic updates (transaction required)
 * - Conflict resolution: Sum all increments (additive)
 * - UUID primary key (offline creation)
 */
export interface ProgressCounter {
  // Identity
  id: string;                    // UUID (client-generated)
  userId: string;                // Foreign key to users table

  // Configuration
  name: string;                  // "Reading Progress", "Running Miles"
  unit: string;                  // "pages", "miles", "times"
  targetValue: number;           // Goal (e.g., 100)
  currentValue: number;          // Current progress

  // Timestamps
  createdAt: string;             // ISO8601
  updatedAt: string;             // ISO8601

  // Sync metadata
  lastSyncedAt?: string;         // ISO8601
  version: number;               // Optimistic locking
  isDeleted: boolean;            // Soft delete
  deletedAt?: string;            // ISO8601
}

/**
 * ProgressCounter creation input
 */
export interface CreateProgressCounterInput {
  name: string;
  unit: string;
  targetValue: number;
}

/**
 * ProgressCounter update input
 */
export interface UpdateProgressCounterInput {
  name?: string;
  currentValue?: number;
  targetValue?: number;
}

File: apps/ios/OYBC/Database/Models/ProgressCounter.swift
import Foundation
import GRDB

/// ProgressCounter - Shared cumulative counter across boards
///
/// Matches TypeScript ProgressCounter interface from @oybc/shared
struct ProgressCounter: Codable, FetchableRecord, PersistableRecord {
    // Identity
    var id: String
    var userId: String

    // Configuration
    var name: String
    var unit: String
    var targetValue: Double
    var currentValue: Double

    // Timestamps
    var createdAt: String // ISO8601
    var updatedAt: String // ISO8601

    // Sync metadata
    var lastSyncedAt: String? // ISO8601
    var version: Int
    var isDeleted: Bool
    var deletedAt: String? // ISO8601

    // MARK: - Database Configuration

    static let databaseTableName = "progress_counters"
}

// MARK: - Computed Properties

extension ProgressCounter {
    /// Computed progress percentage
    var progressPercentage: Double {
        guard targetValue > 0 else { return 0 }
        return (currentValue / targetValue) * 100
    }

    /// Check if target is reached
    var isComplete: Bool {
        return currentValue >= targetValue
    }

    /// Remaining amount to reach target
    var remaining: Double {
        return max(0, targetValue - currentValue)
    }
}
